File: csharp/1013.partition-array-into-three-parts-with-equal-sum.cs
/*
 * @lc app=leetcode id=1013 lang=csharp
 *
 * [1013] Partition Array Into Three Parts With Equal Sum
 */

using System.Linq;

// @lc code=start
public class Solution {
    public bool CanThreePartsEqualSum(int[] A) {
        int sum = A.Sum();
        if (sum % 3 != 0) {
            return false;
        }
        sum /= 3;
        int acc = 0;
        bool hasFirst = false;
        foreach (int num in A.SkipLast(1)) {
            acc += num;
            if (!hasFirst && acc == sum) {
                hasFirst = true;
            } else if (hasFirst && acc == sum * 2) {
                return true;
            }
        }
        return false;
    }
}
// @lc code=end

File: java/1013.partition-array-into-three-parts-with-equal-sum.java
/*
 * @lc app=leetcode id=1013 lang=java
 *
 * [1013] Partition Array Into Three Parts With Equal Sum
 */

// @lc code=start
class Solution {
    public boolean canThreePartsEqualSum(int[] arr) {
        int sum = 0;
        for (int num : arr) {
            sum += num;
        }
        if (sum % 3 != 0) {
            return false;
        }
        int acc = 0, n = arr.length - 1;
        boolean first = false;
        for (int i = 0; i < n; ++i) {
            acc += arr[i];
            if (!first) {
                first = acc == sum / 3;
                continue;
            }
            if (acc == sum / 3 * 2) {
                return true;
            }
        }
        return false;
    }
}
// @lc code=end

